overlord/state/state.go

```diff
@@ -0,0 +1,156 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+// Package state implements the representation of system state.
+package state
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/ubuntu-core/snappy/logger"
+)
+
+// A Backend is used by State to checkpoint on every unlock operation.
+type Backend interface {
+	Checkpoint(data []byte) error
+}
+
+// State represents an evolving system state that persists across restarts.
+//
+// The State is concurrency-safe, and all reads and writes to it must be
+// performed with the state locked. It's a runtime error (panic) to perform
+// operations without it.
+//
+// The state is persisted on every unlock operation via the StateBackend
+// it was initialized with.
+type State struct {
+	// locking
+	mu  sync.Mutex
+	muC int32
+	// storage
+	backend Backend
+	entries map[string]*json.RawMessage
+}
+
+// New returns a new empty state.
+func New(backend Backend) *State {
+	return &State{
+		backend: backend,
+		entries: make(map[string]*json.RawMessage),
+	}
+}
+
+func (s *State) checkpointData() []byte {
+	data, err := json.Marshal(s.entries)
+	if err != nil {
+		// this shouldn't happen, because the actual delicate serializing happens at various Set()s
+		logger.Panicf("internal error: could not marshal state for checkpoiting: %v", err)
+	}
+	return data
+}
+
+// Lock acquires the state lock.
+func (s *State) Lock() {
+	s.mu.Lock()
+	atomic.AddInt32(&s.muC, 1)
+}
+
+// Unlock checkpoint retry parameters (5 mins of retries by default)
+var (
+	UnlockCheckpointMaxRetries    = 10
+	UnlockCheckpointRetryInterval = 30 * time.Second
```

Comment:
Also, given the whole system will be locked down, I'd put something like 1 second here, instead of 30. And change retries to be time bounded, so it's more clear what we mean by that. So, an attempt every 3 seconds for 5 minutes, for example.